lib/inferno/dsl/fhir_evaluation/reference_extractor.rb

```diff
@@ -0,0 +1,78 @@
+module Inferno
+  module DSL
+    module FHIREvaluation
+      class ReferenceExtractor
+        def extract_resource_path_ids(resources)
+          resource_path_ids = Hash.new { |resource_path, id| resource_path[id] = [] }
+
+          resources.each do |resource|
+            resource.each_element do |value, metadata, path|
+              next unless metadata['type'] == 'id'
+              next if path.include?('contained')
+
+              first_path = metadata['path'].partition('.').first.downcase
+              resource_path_ids[first_path] << value
+            end
+          end
+
+          resource_path_ids
+        end
+
+        def extract_references(resources, resource_path_ids)
+          @references = Hash.new { |reference, id| reference[id] = [] }
+          @resource_path_ids = resource_path_ids
+
+          resources.each do |resource|
+            extract_references_resource(resource)
+          end
+
+          @references
+        end
+
+        def extract_references_resource(resource)
```

Comment:
This name does not make sense to me. What is the relationship between `references` and `resource`?

Reply:
Not sure what would be a good naming for that, the function was created to address Rubocop allergic to Cyclomatic complexity:

`Metrics/CyclomaticComplexity: Cyclomatic complexity for extract_references is too high`

Reply:
What does the method do?

Reply:
Originally `extract_references` and `extract_references_resource` were one. I separated them merely to avoid cyclomatic complexity. 
They may be something like `extract_references_level_one` and `extract_references_level_two`...

Reply:
If the point is to extract the references in a single resource, I'd suggest just something like `extract_references_from_resource`

Reply:
Good suggestion. Modified https://github.com/inferno-framework/inferno-core/pull/583/commits/7ec179865dfc197fbdccdbc6642320285ed5989c